### ISSUE TITLE
argument.type.incompatible error when combining "? super" and visitor pattern

### ISSUE THREAD

**Issue by sebright** [typetools/checker-framework#1712]
I am trying to run the Checker Framework Nullness Checker on https://github.com/census-instrumentation/opencensus-java, with Checker Framework version 2.3.0.  I saw many unexpected errors related to the use of the visitor pattern in classes like [AggregationData](https://github.com/census-instrumentation/opencensus-java/blob/5d772db096ecf4a596e0bfd93dce072d5cfc0e80/api/src/main/java/io/opencensus/stats/AggregationData.java#L47).

I was able to reproduce the error with the live demo, with the following code:

```java
import org.checkerframework.checker.nullness.qual.Nullable;

abstract class Superclass {

  abstract <T> T match(
      Function<? super SubclassA, T> visitA, Function<? super SubclassB, T> visitB);

  class SubclassA extends Superclass {
    @Override
    <T> T match(Function<? super SubclassA, T> visitA, Function<? super SubclassB, T> visitB) {
      return visitA.apply(this);  // line 11
    }
  }

  class SubclassB extends Superclass {
    @Override
    <T> T match(Function<? super SubclassA, T> visitA, Function<? super SubclassB, T> visitB) {
      return visitB.apply(this);  // line 18
    }
  }

  abstract class Function<T1, T2> {
    abstract T2 apply(T1 arg);
  }
}
```
There were errors at both calls to `apply`:

line 11:
```
Error: [argument.type.incompatible] incompatible types in argument.
   found   : /*DECL*/ /*DECL*/ @UnknownKeyFor Superclass.@UnknownKeyFor SubclassA
   required: ?[ extends @UnknownKeyFor Object super @KeyForBottom Superclass.@KeyForBottom SubclassA]
```

line 18:
```
Error: [argument.type.incompatible] incompatible types in argument.
   found   : /*DECL*/ /*DECL*/ @UnknownKeyFor Superclass.@UnknownKeyFor SubclassB
   required: ?[ extends @UnknownKeyFor Object super @KeyForBottom Superclass.@KeyForBottom SubclassB]
```

I expected the types in the warnings to be compatible, given that UnknownKeyFor is at the top of the type hierarchy and KeyForBottom is at the bottom.  I don't understand the Checker Framework type system very well, though, so I could be wrong.

The check passed when I removed all uses of "? super".

Screenshot:

![checker-framework-demo-2](https://user-images.githubusercontent.com/19807615/34066771-6f16096e-e1c9-11e7-8f33-f55c38540f13.png)

**Comment by sebright** [comment]
Thanks!
